GH-33892: [R] Map `dplyr::n()` to `count_all` kernel

ianmcook opened the pull request:
### Rationale for this change

This PR is a follow-up to #15083. It allows the R package to register bindings to nullary aggregation functions, and it remaps `dplyr::n()` to the nullary aggregation function `count_all`.

This PR also:
- Prepares the R bindings to support aggregation functions with 2+ arguments, although none yet exist in the C++ library
- Removes the heuristics that were used to infer the data types of aggregates, replacing that with actual type determination

### Are these changes tested?

Yes, through existing tests.

### Are there any user-facing changes?

No.
* Closes: #33892
* Closes: #33960

ianmcook commented:
I'm seeing an ~8% speed improvement when I run this locally on my Intel Mac:
```r
library(arrow)
library(dplyr)
library(tictoc)

z <- arrow_table(z = seq_len(4e9))

tic()
z %>% summarize(n()) %>% collect()
toc()
```
I am having trouble getting the package to build with Dataset so that's about the largest data I can test on at the moment.

nealrichardson commented:
I'll try to take a look this week.

nealrichardson commented on r/R/dplyr-summarize.R:
```diff
@@ -322,15 +301,76 @@ arrow_eval_or_stop <- function(expr, mask) {
+  imap(
+    fields,
+    ~if (hash && aggs[[.y]]$fun == "tdigest") {
+        fixed_size_list_of(float64(), 1L)
```
Is it always float64()? (looks like it is today, but probably not safe to assume that things like timestamp wouldn't return timestamp)

Is it always length 1? (looks like not, the length depends on the number of quantiles you give it in the options)

I think this would generalize better:

```suggestion
        fixed_size_list_of(.x, length(aggs[[.y]][["options"]][["q"]])
```

nealrichardson replied:
Nevermind, see above, I think we can construct `hash_` Expressions and get the actual output type without any munging.

nealrichardson commented on r/R/dplyr-summarize.R:
```diff
@@ -322,15 +301,76 @@ arrow_eval_or_stop <- function(expr, mask) {
+}
+
+# Unlike with other pairs of non-hash/hash aggregate kernels in the Arrow C++
+# library, the `tdigest` and `hash_tdigest` kernels have different output types.
```
Perhaps a safer way to do the type determination is to build expressions with the hash versions. The prose documentation says the hash kernels are unary, but according to the error messages you get, as well as the inline function docs in hash_aggregate.cc, they're binary. After some trial and error, I got this to work:

```
> x <- Expression$create("tdigest", Expression$field_ref("a"), options=list(q=c(.1, .2)))
> x$type(schema(a = int32()))
Float64
double

> x <- Expression$create("hash_tdigest", Expression$field_ref("a"), Expression$scalar(1L)$cast(uint32()), options=list(q=c(.1, .2)))
> x$type(schema(a = int32()))
FixedSizeListType
fixed_size_list<item: double>[2]
```

So the second argument is the group hash, and that is `uint32()`.

ianmcook replied:
Very cool! 

I changed it to work like this, but I used `Scalar$create(1L, uint32())` instead of `Expression$scalar(1L)$cast(uint32())` as the groups argument.

nealrichardson commented on r/R/dplyr-summarize.R:
```diff
@@ -322,15 +301,76 @@ arrow_eval_or_stop <- function(expr, mask) {
+aggregated_fields <- function(aggs) {
+  map(
+    aggs,
+    ~Expression$create(.$fun, args = .$data, options = .$options)
```
See the comment below. Basically, `if (hash)` then it's `paste0("hash_", .$fun)` and `c(.$data, Expression$scalar(Scalar$create(1L, type = uint32())))`

nealrichardson commented on r/R/dplyr-summarize.R:
```diff
@@ -322,15 +301,76 @@ arrow_eval_or_stop <- function(expr, mask) {
+
+# This function returns a list of expressions representing the aggregated fields
+# that will be returned by an aggregation
+aggregated_fields <- function(aggs) {
```
Technically these aren't "Fields". Perhaps `aggregation_exprs` instead?

nealrichardson commented on r/R/query-engine.R:
```diff
@@ -101,7 +101,11 @@ ExecPlan <- R6Class("ExecPlan",
         # dplyr does not error on this but the result it gives isn't great
-        node <- node$Project(summarize_projection(.data))
+        projection <- summarize_projection(.data)
+        # skip projection if no grouping and all aggregate functions are nullary
```
Why?

ianmcook replied:
In general we want to avoid adding unnecessary projections to the ExecPlan. This is a case where we can do that.

nealrichardson replied:
Would something be harmed if it were projected to having 0 columns? Would that be even more efficient?

ianmcook replied:
My understanding is that adding a project node before an aggregate node will never improve performance, and that the only real reason to do it is to ensure that the aggregate node has all the columns it needs.

I just checked with Weston about this. Here's our conversation:

**Ian:**
Which of these two ExecPlans would execute the most quickly and efficiently?

1. Read source data containing columns A,B,C,D --> project to only columns A and B --> aggregate sum of A over groups in B --> output to table

2. Read source data containing columns A,B,C,D --> aggregate sum of A over groups in B --> output to table

In other words, does adding an unnecessary projection help, hurt, or not make any difference?

**Weston**:
I assume by "project to" you mean an actual project node and not a push down projection? In reality you'd want to avoid reading C and D from disk so you'd want the projection in the read.

**Ian**:
correct, actual project node

**Weston**:
An actual project node, with just field references, is effectively free.

Now, if the follow-up node was a filter node, then you might be paying more for plan 2.  We don't use selection vectors so the filter node would have to materialize the data and it would have to materialize your two junk columns.

The aggregate node, on the other hand, doesn't output any input columns.  So I'd expect it will throw them away in it's InputReceived and the performance of 1 & 2 will be more or less the same.

nealrichardson commented on r/R/dplyr-summarize.R:
```diff
@@ -322,15 +301,76 @@ arrow_eval_or_stop <- function(expr, mask) {
+
+# This function determines what names to give to the fields used in aggregations
+# (the "targets"). When an aggregate function takes 2 or more fields as targets,
+# this function gives the fields unique names by appending `..1`, `..2`, etc.
```
Why? I'm not sure about this. Since there's no code that actually does this in this PR, maybe save it for #33959, where you can prove out what works/is necessary?

ianmcook replied:
I plan to do a PR for #33959 soon. I would prefer to leave this code in and change it as needed in that PR. In the meantime, it is unreachable code, which I know isn't great. But taking it out leaves things in an awkward state where obvious cases are unhandled in the code.

nealrichardson commented on r/R/dplyr-summarize.R:
```diff
@@ -322,15 +301,76 @@ arrow_eval_or_stop <- function(expr, mask) {
   out
 }
 
+# This function returns a list of expressions that can be used to project the
```
```suggestion
# This function returns a list of expressions that is used to project the
```

nealrichardson reviewed (commented):
Thanks for taking this on. Here are some suggestions. I think in the end this PR will result in simplifying the code a bit, it should be more understandable what's going on and will remove some hacks/imprecision.

ianmcook commented:
@nealrichardson thanks for your review! I think I addressed or responded to all the issues you raised. I also changed some of the helper functions so that they return types instead of Expressions. They are only ever used to assemble the schema, so it is clearer this way.

nealrichardson commented on r/R/dplyr-summarize.R:
```diff
@@ -322,15 +301,71 @@ arrow_eval_or_stop <- function(expr, mask) {
+
+# This function returns a named list of the data types of the aggregate columns
+# returned by an aggregation
+aggregate_types <- function(.data, hash, schema = NULL) {
```
Is schema needed as an argument here? The Expressions in $data will all bring schema with them.

ianmcook replied:
The code in `dplyr-collect.R` passes `old_schm` to this function. That's how it worked before and I didn't want to change in for fear of breaking it.

nealrichardson replied:
That could predate the schema tracking in Expressions. I'd try removing it, and if the tests pass, I guess it's not needed? After all, that's what the tests are there to tell us.

ianmcook replied:
I removed it and it looks like the tests pass.

ianmcook replied:
Oops, I take that back. It looks like there is an edge case with datasets where the old schema does need to be passed, and this did cause a test to fail: https://github.com/apache/arrow/actions/runs/4147404413/jobs/7174247433#step:5:24083

ianmcook replied:
I reverted the removal and restored the schema argument here.

nealrichardson commented on r/R/dplyr-summarize.R:
```diff
@@ -322,15 +301,71 @@ arrow_eval_or_stop <- function(expr, mask) {
+        # hash aggregate kernels must be passed another argument representing
+        # the groups, so we pass in a dummy scalar, since the groups will not
+        # affect the type that an aggregation returns
+        args = c(.$data, Scalar$create(1L, uint32())),
```
Minor optimization: we could define `.hash_arg <- Scalar$create(1L, uint32())` outside the function so we don't have to keep doing it. There's small (<1ms) overhead in creating R6 objects, and this creates 2 every time you call it. When I was doing some benchmarking of the query building a while back, these all added up such that our conbench TPC-H query benchmarks were being driven more by them than evaluating the query itself (since they're not using really big data).

ianmcook replied:
Defining it outside of a function doesn't work because of build order. So I did the next best thing and pulled the definition outside of the `map()` so it at least it only incurs the overhead once per function call.

nealrichardson reviewed (commented):
A couple of small suggestions but otherwise LGTM, thanks for taking this on!

ianmcook commented on r/R/dplyr-collect.R:
```diff
@@ -179,18 +179,18 @@ implicit_schema <- function(.data) {
+    hash <- length(.data$group_by_vars) > 0
     # The output schema is based on the aggregations and any group_by vars
-    new_fields <- map(summarize_projection(.data), ~ .$type(old_schm))
+    new_fields <- c(
+      aggregate_types(.data, hash, old_schm),
+      group_types(.data, old_schm)
+    )
```
Indeed. Thanks for noticing this.
